### PR TITLE
Add caching for theme yaml file

### PR DESCRIPTION
**Pull request by jocel1**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | Add caching for theme yaml file to speed up load time
| Type?         | improvement
| Category?     | BO
| BC breaks?    | no
| Deprecations? | no
| How to test?  | Load a product page

### CODE REVIEW

**Comment by jocel1** [comment]
https://blackfire.io/profiles/compare/dd0d8638-6df8-4f40-b0ca-c7726774bb7c/graph

File: src/Core/Addon/Theme/ThemeRepository.php [hunk 1]
```diff
@@ -52,18 +52,22 @@ public function __construct(ConfigurationInterface $configuration, Filesystem $f
...
+            $jsonConf = $themeCachePath . '/shop' . $this->shop->id . '.json';
+        } else {
+            $jsonConf = $themeCachePath . '/theme.json';
         }
 
         if ($this->filesystem->exists($jsonConf)) {
             $data = $this->getConfigFromFile($jsonConf);
         } else {
             $data = $this->getConfigFromFile($dir.'/config/theme.yml');
+            file_put_contents($jsonConf, json_encode($data));
```
Comment (Quetzacoalt91) [inline src/Core/Addon/Theme/ThemeRepository.php]: Can you please use the `dumpFile()` function from the class [Filesystem](https://symfony.com/doc/2.8/components/filesystem.html#dumpfile)?
The instance is available in `$this->filesystem`.

**Comment by Quetzacoalt91** [comment]
So you're telling me the cache was never created, although we always checked it?

**Comment by jocel1** [comment]
Only if there's no preexisting shop-specific json files. (it occurs after an upgrade for example)
The files are only created when ThemeManager->saveTheme() is called

**Comment by Quetzacoalt91** [comment]
I'm wondering if the other unlink/dump occurrences of these Json files are useful, because they will be created by this getInstanceByName() anyway.

By the way, could we add a check with the file timestamps? This would allow us to regenerate the json file as soon as the YML file is modified. And we can also check the environment in order to to this only in dev mode.
I'm open to any discussion, because this suggestion may break something. I do not own this part of the project. :grimacing:

**Comment by jocel1** [comment]
right about the cache invalidation. I can modify as well saveTheme() to just invalidate the cache instead of regenerating it.

**Comment by jocel1** [comment]
After reading again the code, I don't think my fix introduce any potential regression. 
The old "normal" behaviour is to generate the json cache file once the theme is saved (through the ThemeManager->saveTheme() call). Once the cache file is generated, it's never invalidated.

However, I need to check why the saveTheme() is not called after a migration, the real fix which needs to be done is perhaps here.

**Comment by aleeks** [comment]
Hello @jocel1 
It's not really a bug fix so i move this PR on `develop`branch, you need to rebase from develop!
Thank you!

**Comment by Quetzacoalt91** [comment]
> Once the cache file is generated, it's never invalidated.

Yeah, and it seems it could be better here. Once you import a theme, the data modified in the yml config file is never taken in account, which keeps stale data loaded during the development.

**Comment by jocel1** [comment]
Actually I have checked, and the issue only occurs if the upgrade is run in command line without a theme upgrade. The case in this PR will probably never happen for real. But yet it could be useful to create a JIRA to track the cache invalidation issue

**Review by Quetzacoalt91 — APPROVED** [review]

**Comment by Quetzacoalt91** [comment]
Thank you @jocel1

**Comment by jocel1** [comment]
Actually the case in this PR occurs for real, when the cache is purged, or the codebase is redeployed without any cache files. In this case the theme cache is never regenerated, and the application performance is poor (I've encounter the case with a real customer). 
@eternoendless It could perhaps make sense to merge it in 1.7.2.x since it's a bug.

**Comment by eternoendless** [comment]
There will be no more releases for 1.7.2.x except for critical security fixes.

**Comment by kpodemski** [comment]
That's a shame, if we clear the cache, we loose columns settings, also undefined variables during AJAX requests, these are critical bugs...

**Comment by eternoendless** [comment]
@kpodemski we can discuss about it on Gitter if you want to

**Pull request merged**